nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -0,0 +1,224 @@
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  action_server_->deactivate();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+WaypointFollower::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  action_server_.reset();
+  nav_to_pose_client_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+WaypointFollower::on_error(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_FATAL(get_logger(), "Lifecycle node entered error state");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+WaypointFollower::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void
+WaypointFollower::followWaypoints()
+{
+  auto goal = action_server_->get_current_goal();
+  auto feedback = std::make_shared<ActionT::Feedback>();
+  auto result = std::make_shared<ActionT::Result>();
+
+  // Check if request is valid
+  if (!action_server_ || !action_server_->is_server_active()) {
+    RCLCPP_DEBUG(get_logger(), "Action server inactive. Stopping.");
+    return;
+  }
+
+  rclcpp::Rate r(10);
+  uint goal_index = 0;
+  bool new_goal = true;
+
+  while (rclcpp::ok()) {
+    // Check if asked to stop processing action
+    if (action_server_->is_cancel_requested()) {
+      RCLCPP_INFO(get_logger(), "Cancelling action.");
+      action_server_->terminate_goals();
+      return;
+    }
+
+    // Check if asked to process another action
+    if (action_server_->is_preempt_requested()) {
+      RCLCPP_INFO(get_logger(), "Preempting the goal pose.");
+      goal = action_server_->accept_pending_goal();
+      goal_index = 0;
+      new_goal = true;
+    }
+
+    // Check if we need to send a new goal
+    if (new_goal) {
+      new_goal = false;
+      ClientT::Goal client_goal;
+      client_goal.pose = goal->poses[goal_index];
+
+      auto send_goal_options = rclcpp_action::Client<ClientT>::SendGoalOptions();
+      send_goal_options.result_callback =
+        std::bind(&WaypointFollower::resultCallback, this, std::placeholders::_1);
+      send_goal_options.goal_response_callback =
+        std::bind(&WaypointFollower::goalResponseCallback, this, std::placeholders::_1);
+      auto future_goal_handle =
+        nav_to_pose_client_->async_send_goal(client_goal, send_goal_options);
+      current_goal_status_ = ActionStatus::PROCESSING;
+    }
+
+    feedback->current_waypoint = goal_index;
+    action_server_->publish_feedback(feedback);
+
+    if (current_goal_status_ == ActionStatus::FAILED) {
+      if (stop_on_failure_) {
+        RCLCPP_WARN(get_logger(), "Failed to process waypoint %i in waypoint "
+          "list and stop on failure is enabled."
+          " Terminating action.", goal_index);
+        action_server_->terminate_goals();
+        return;
+      } else {
```

Comment:
Since this is the interface to navigation, I think a retry here messes with the encapsulation of tasks.

I think of this like "I just asked Navigation to go somewhere and it said it couldn't do it, time to do the next thing". If there was a retry to be done, that should be considered in the layer below this in the navigation task to retry until unable to continue.

Reply:
OK, well you have a "stop_on_failure" but if that is "false" you are going to the next waypoint. That has me slightly worried as it could result in cutting a corner that shouldn't be cut. That's why I suggested a retry instead.

Reply:
I think if its important for a navigation task to try really hard to complete a task, they should turn the retry knob in navigation itself. I don't think we should rely on the layers above for that, I think we need to trust that the navigation task knows what its doing and returning a failure means it really failed. 

You'd have to describe what you mean by cutting a corner. You shouldn't hit anything around a corner since the next waypoint is still a navigation task and navigation shouldn't run into things, but you will miss that single waypoint

Reply:
That is what I mean by cutting a corner, missing the waypoint. The example I think of is an autonomous tractor plowing a field. The waypoints are the ends of the field where it has to do a U-turn to come back. If it gets halfway to the other side, then for some reason fails to get to the end, I don't think you'd want it turning to go to the next waypoint. But I also see your point that we have a retry in the behavior tree, so maybe retry is wrong too. Maybe it's best that it just stops on failure always, and punts the decision what to do next back to the action client, because any assumptions made at this level could be invalid.

Reply:
> he waypoints are the ends of the field where it has to do a U-turn to come back. If it gets halfway to the other side, then for some reason fails to get to the end, I don't think you'd want it turning to go to the next waypoint.

See, I would think you would want that. We don't know (currently) why it failed, so I'd rather go to the next one than sit there potentially forever because something's blocking the way. Its also a parameter to set if you want to exit the task on any waypoint failure, and I think that's the logical conclusion to draw if its not acceptable to the program to miss a single point. 

> But I also see your point that we have a retry in the behavior tree, so maybe retry is wrong too.

I just wouldn't want to break the encapsulation of tasks, navigation has the ability to retry, so adding it here gives people 2 identical knobs and its easy to get messed up about what's happening. For tasks where you **really** want to try hard to get to a waypoint but want to continue moving on, then the retry in navigation should be high and have good recovery behaviors to try to get there. If they can't and returns failure, the client should trust that by the way the user has created navigation that a failure is truly a failure.

> Maybe it's best that it just stops on failure always

The option is there with `stop_on_failure`. But I have to say from experience that would be really bad to just stop on any failure. There's lots of failures with robots in dynamic spaces because people suck and things move day to day. I'd rather my robot do 90% of the work and I can leave it alone for a day than have to get emails every few hours that its stuck for me to deal with.

Reply:
I think you have a different use case in mind than I do. The only ones I can think of, I think I'd want to handle the failure in the client logic, but I won't die on this hill. If you have a use case in mind that would work better by continuing to the next waypoint, I'd like to understand what it is, but I won't block this PR regardless, as the `stop_on_failure` is defaulting to `True` which covers the ones I know of.